logging/src/lib.rs

```diff
@@ -0,0 +1,340 @@
+    /// Communication channel identifier
+    pub comm_channel: Option<usize>,
+    /// Source worker index.
+    pub source: usize,
+    /// Target worker index.
+    pub target: usize,
+    /// Message sequence number.
+    pub seq_no: usize,
+    /// Number of typed records in the message.
+    pub length: usize,
+}
+
+/// Records the starting and stopping of an operator.
+#[derive(Abomonation, Debug, Clone, PartialEq, Eq)]
+pub enum StartStop {
+    /// Operator starts.
+    Start,
+    /// Operator stops; did it have any activity?
+    Stop {
+        /// Did the operator perform non-trivial work.
+        activity: bool
+    },
+}
+
+#[derive(Abomonation, Debug, Clone)]
+/// Operator start or stop.
+pub struct ScheduleEvent {
+    /// Worker-unique identifier for the operator, linkable to the identifiers in `OperatesEvent`.
+    pub id: usize,
+    /// `Start` if the operator is starting, `Stop` if it is stopping.
+    /// activity is true if it looks like some useful work was performed during this call (data was
+    /// read or written, notifications were requested / delivered)
+    pub start_stop: StartStop,
+}
+
+#[derive(Abomonation, Debug, Clone)]
+/// Application-defined code startor stop
+pub struct ApplicationEvent {
+    /// Unique event type identifier
+    pub id: usize,
+    /// True when activity begins, false when it stops
+    pub is_start: bool,
+}
+
+#[derive(Abomonation, Debug, Clone)]
+/// Application-defined code start or stop
+pub struct GuardedMessageEvent {
+    /// True when activity begins, false when it stops
+    pub is_start: bool,
+}
+
+#[derive(Abomonation, Debug, Clone)]
+/// Application-defined code start or stop
+pub struct GuardedProgressEvent {
+    /// True when activity begins, false when it stops
+    pub is_start: bool,
+}
+
+#[derive(Abomonation, Debug, PartialEq, Eq, Hash, Clone, Copy)]
+pub struct EventsSetup {
+    pub index: usize,
+}
+
+#[derive(Abomonation, Debug, PartialEq, Eq, Hash, Clone, Copy)]
+pub struct CommsSetup {
+    pub sender: bool,
+    pub process: usize,
+    pub remote: Option<usize>,
+}
+
+#[derive(Abomonation, Debug, Clone)]
+pub enum CommChannelKind {
+    Progress,
+    Data,
+}
+
+#[derive(Abomonation, Debug, Clone)]
+pub struct CommChannelsEvent {
+    pub comm_channel: Option<usize>,
+    pub comm_channel_kind: CommChannelKind,
+}
+
+#[derive(Abomonation, Debug, Clone)]
+pub struct InputEvent {
+    pub start_stop: StartStop,
+}
+
+#[derive(Debug, Clone, Abomonation)]
+pub enum Event {
+    /*  0 */ Operates(OperatesEvent),
+    /*  1 */ Channels(ChannelsEvent),
+    /*  2 */ Progress(ProgressEvent),
+    /*  3 */ PushProgress(PushProgressEvent),
+    /*  4 */ Messages(MessagesEvent),
+    /*  5 */ Schedule(ScheduleEvent),
+    /*  6 */ Application(ApplicationEvent),
+    /*  7 */ GuardedMessage(GuardedMessageEvent),
+    /*  8 */ GuardedProgress(GuardedProgressEvent),
+    /*  9 */ CommChannels(CommChannelsEvent),
+    /* 10 */ Input(InputEvent),
+}
```

Comment:
We've talked a bit about this before, and I'm still not in love with it yet. :) If we treat all of the logging stuff as unstable, then no worries because we can try things out, but I think:

1. We may want separate streams, so that individual streams can be selectively (and programmatically) enabled for capture. E.g. right now, I could imagine wanting to turn off the `Schedule` events because they are noisy af.

2. We may also want to support user-defined logging, at least playing out the fantasies I had with ETW back at MSR. :) If a user has higher-level begin-end events that they want to track, we probably wouldn't want to require a monolithic enum that they extend (perhaps we don't, and this enum gets locked down; no worries then).

Anyhow, just writing these down rather than putting them up as a roadblock!